[core] Return 405 instead of 404 when applicable

HexRealm opened the pull request:
This pull request implements the feature requested in the issue #228 and adds a new test case. I've made the feature toggable and disabled by default to preserve backwards compatibility. I'm open to any suggestions on how this can be improved.

majorpasza commented on src/main/java/io/javalin/core/JavalinServlet.kt:
```diff
@@ -74,7 +75,44 @@ class JavalinServlet(
                 }
             }
             if (endpointEntries.isEmpty() && type != HandlerType.HEAD && type != HandlerType.GET) {
```
I did not study the code thoroughly, but I think it would be better to throw HaltException(405) (if possible) here instead of rendering 405 as if it would be normal. That is because I think that exception mapping won't be triggered for this https://javalin.io/documentation#exception-mapping

Here I mentioned that there might be couple of places in Javalin that are hardcoded, which may be undersirable given some strict requirements, this would be one of them.
https://github.com/tipsy/javalin/issues/219

majorpasza commented on src/main/java/io/javalin/core/JavalinServlet.kt:
```diff
@@ -74,7 +75,44 @@ class JavalinServlet(
+                                </body>
+                            </html>
+                        """
+                    ctx.status(405)
```
I did not study the code thoroughly, but I think it would be better to throw HaltException(405) (if possible) here instead of rendering 405 as if it would be normal. That is because I think that exception mapping won't be triggered for this https://javalin.io/documentation#exception-mapping

Here I mentioned that there might be couple of places in Javalin that are hardcoded, which may be undersirable given some strict requirements, this would be one of them.
https://github.com/tipsy/javalin/issues/219

HexRealm commented:
I've updated the PR with some of the changes requested, I was thinking if I should be checking if there's any objectToJson available before trying to create the json response, because if there isn't i'm not sure how that kotlin code will behave.

majorpasza commented on src/main/java/io/javalin/core/util/MethodNotAllowedUtil.kt:
```diff
@@ -0,0 +1,47 @@
+object MethodNotAllowedUtil {
+
+    @JvmStatic
+    fun findAvailableHandlerTypes(matcher: PathMatcher, requestUri: String): List<HandlerType> {
```
```
enumValues<HandlerType>().flatMap { matcher.findEntries(it, requestUri)  }
``` 
Maybe I'm biased, but this looks more readable to me

ShikaSD replied:
You can also use an extension function on `PathMatcher` instead of creating a util
```kotlin
fun PathMatcher.findAvailableHandlerTypes(requestUri: String): List<HandlerType>
```

Then it can be used in `JavalinServlet` like this:
```kotlin
val availableHandlerTypes = matcher.findAvailableHandlerTypes(requestUri)
```

majorpasza commented:
@HexRealm 
From what I remember, if you try to serialize some object and objectmapper is not on classpath, exception is thrown.
But, in this case if objectToJson is null , I'd throw HaltException(405, body) with html body.

ShikaSD commented on src/main/java/io/javalin/core/JavalinServlet.kt:
```diff
@@ -74,7 +77,25 @@ class JavalinServlet(
+
+                    val acceptableReturnTypes = ctx.header("Accept")
+
+                    val body = if (acceptableReturnTypes != null && acceptableReturnTypes.contains("application/json")) {
```
What if acceptable return type is `application/yaml`?
Maybe it is a better idea to just send a `HaltException` with simple text by default?
Then user can override it in exception mapper or so.

david-wg2 replied:
I think making special cases for HTML and JSON is okay, they will account for something like 99.9% of all requests. A general string formatted as JSON would also work for everything though.

ShikaSD commented on src/main/java/io/javalin/core/util/MethodNotAllowedUtil.kt:
```diff
@@ -0,0 +1,47 @@
+    fun findAvailableHandlerTypes(matcher: PathMatcher, requestUri: String): List<HandlerType> {
+        val availableHandlerTypes = ArrayList<HandlerType>()
+
+        enumValues<HandlerType>().forEach {
```
This will get all the handlers including `before`, `ws` and `after`.
I am not sure that the 405 status should be returned when only `before` or `after` handler is set, as these are not actual HTTP methods.

david-wg2 replied:
You're absolutely correct, before and after are not applicable. WebSocket technically is, but I'm not sure if it should be included either.

ShikaSD commented on src/main/java/io/javalin/core/JavalinServlet.kt:
```diff
@@ -74,7 +77,25 @@ class JavalinServlet(
             if (endpointEntries.isEmpty() && type != HandlerType.HEAD && type != HandlerType.GET) {
-                throw HaltException(404, "Not found")
+
+                if (prefer405over404) {
```
The body of the condition can be moved to the private function to prevent bloating the method with logic.

ShikaSD commented on src/main/java/io/javalin/core/JavalinServlet.kt:
```diff
@@ -74,7 +77,25 @@ class JavalinServlet(
-                throw HaltException(404, "Not found")
+
+                if (prefer405over404) {
+                    val availableHandlerTypes = MethodNotAllowedUtil.findAvailableHandlerTypes(matcher, requestUri)
```
Calling this here will do matcher work twice: first for called `HandlerType`, then second time here.
Consider searching for all the handlers matching uri in the beginning of the method and then reusing that pool for here.

david-wg2 replied:
The performance penalty is only applicable for 404s (well, 405s), so it doesn't really matter. I prefer doing the work twice since it keeps the functionality isolated.

david-wg2 commented:
Sorry about the slow progress on this @HexRealm. I'm still travelling, and will be back in the middle of next week.

> @HexRealm
> From what I remember, if you try to serialize some object and objectmapper is not on classpath / not specified, exception is thrown and it results in status code 500.
But, in this case if objectToJson is null , I'd throw HaltException(405, body) with html body, unless the caller specified he expects application/json explicitly, in this case we have 2 options:
code 500 as server configuration is unable to respond to the request because it is misconfigured and it is servers fault.
Halt(405, body) where body is empty json.
@tipsy ?

We shouldn't use the objectmapper at all, just create JSON with a template string.

HexRealm commented:
I've updated the PR with the following changes:

- Moved the method not allowed logic to a private function
- Use a string template instead of the objectmapper for json responses
- Updated the test case so it covers multiple handlers available instead of only one

HexRealm commented:
I've also just added a change so `findAvailableHandlerTypes` is now called `findAvailableHttpHandlerTypes` and it'll only return true http `HandlerType`s. To accomplish this i've added a `isHttpMethod` to the `HandlerType` enum with the intent of providing anyone a way to check if the handlerType is a true http method or not.

david-wg2 commented on src/main/java/io/javalin/core/HandlerType.kt:
```diff
@@ -11,6 +11,13 @@ import javax.servlet.http.HttpServletRequest
 enum class HandlerType {
     GET, POST, PUT, PATCH, DELETE, HEAD, TRACE, CONNECT, OPTIONS, BEFORE, AFTER, INVALID, WEBSOCKET;
 
+    fun isHttpMethod(): Boolean {
```
You can write this as a function expression: 

```
fun isHttpMethod() = when (this) {
```

david-wg2 commented on src/main/java/io/javalin/core/JavalinServlet.kt:
```diff
@@ -74,7 +76,12 @@ class JavalinServlet(
             if (endpointEntries.isEmpty() && type != HandlerType.HEAD && type != HandlerType.GET) {
-                throw HaltException(404, "Not found")
+
+                if (prefer405over404) {
```
I think this if/else section would read better if it looked more like this:

```java
if (prefer405over404 && otherMethodsAvailable()) {
    throw HaltException(405, getAvailableMethods());
}
throw HaltException(404, "Not found")
```

HexRealm commented:
I've updated the PR trying to take in consideration what @tipsy said in regards to the code readability and what @ShikaSD said earlier about organization and performance.

tipsy commented:
Thanks @HexRealm, good work! Sorry about the delay, I just got back from travelling today. 

I think we'll include this in v2 (probably after during/after summer). Figuring things out now!

tipsy commented:
This will be part of 2.0.0.RC0 @HexRealm, but hopefully that won't take too long :)

HexRealm commented:
Thank you all for the reviews & suggestions
